src/sisyphus/task.clj

```diff
@@ -171,27 +177,136 @@
       (log/notice! "STEP FAILED" (:workflow task) (:name task) log)
       (status!
        kafka task "step-error"
-       {:code code
+       {:success success?
         :log log}))))
 
+(defn make-timer
+  "Make a `future` as a timer that calls `f` in another thread. Cancelling it
+  can stop the timer but won't interrupt `f` midway."
+  [milliseconds f]
+  (future
+    (Thread/sleep milliseconds)
+    (future (f))))
+
+(defn cancel-timer
+  "Cancel a timer if possible. nil safe. Return true if this cancelled it."
+  [timer]
+  (and timer (future-cancel timer)))
+
+(defn- terminating-state
+  "Return the new state-map for an attempted state transition from :running
+  to/towards a termination status. If the task id matches, then status goes
+  :starting -> :terminate-when-ready, or :running -> terminating-status, or no
+  change. A transition to :action = :kill means it's time to kill the process;
+  another value is the termination status when the process is ready to kill."
+  [state-map task-id termination-status]
+  (cond
+    (not (and task-id (= task-id (:id (:task state-map)))))
+    state-map
+
+    (= (:status state-map) :starting)
+    (assoc state-map :status :terminate-when-ready :action termination-status)
+
+    (= (:status state-map) :running)
+    (assoc state-map :status termination-status :action :kill)
+
+    :else
+    state-map))
+
+(defn- terminate!
+  "Terminate the task's docker process if it's running; setting status to track
+  how it finished and prevent double-termination. Call this in a forked thread.
+  (perform-task! will finish up soon.)"
+  [{:keys [docker kafka state]} task-id termination-status]
+  (let [[old new] (swap-vals! state terminating-state task-id termination-status)
+        task (:task new)
+        docker-id (:docker-id new)]
+    (when (and (= (:action new) :kill) (not= (:action old) :kill))
+      (log/debug! "terminating step..." termination-status)
+      (docker/stop! docker docker-id)
+      (log/notice! "STEP TERMINATED" termination-status)
+      (status! kafka task "step-terminated" {:reason termination-status})
+      true)))
+
+(defn- terminate-by-timeout!
+  [sisy-state task-id]
+  (terminate! sisy-state task-id :terminated-by-timeout))
+
+(defn terminate-by-request!
+  [sisy-state task-id]
+  (terminate! sisy-state task-id :terminated-by-request))
```

Comment:
This is becoming a serious state machine! Something like this could be useful, but maybe overkill? https://github.com/ztellman/automat

Reply:
Maybe this could be simpler. It's synchronizing 3 threads that are collectively controlling docker and the command itself running in external processes. Or maybe it needs a concise FSM diagram.